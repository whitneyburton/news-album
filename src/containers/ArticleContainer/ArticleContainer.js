import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Article } from '../../components/Article/Article';
import Masonry from 'react-masonry-css';

export const ArticleContainer = ({
  match,
  isDisabled,
  natGeoArticles,
  cryptoCoinsArticles,
  newScientistArticles
}) => {
  const generateArticleCategory = () => {
    switch (match.path) {
      case '/national-geographic':
      case '/national-geographic/:id':
        return natGeoArticles.map(article => {
          return <Article key={article.id} article={article} match={match} />;
        });
      case '/crypto-coins':
      case '/crypto-coins/:id':
        return cryptoCoinsArticles.map(article => {
          return <Article key={article.id} article={article} match={match} />;
        });
      case '/new-scientist':
      case '/new-scientist/:id':
        return newScientistArticles.map(article => {
          return <Article key={article.id} article={article} match={match} />;
        });
      case '/favorites':
      case '/favorites/:id':
        const allArticles = [
          ...natGeoArticles,
          ...cryptoCoinsArticles,
          ...newScientistArticles
        ];
        return generateFavoriteArticles(allArticles, match);
      default:
        return;
    }
  };

  const generateFavoriteArticles = (allArticles, match) => {
    const favoriteTitles = JSON.parse(localStorage.getItem('favorites')) || [];
    const favoriteArticles = allArticles.reduce((acc, article) => {
      favoriteTitles.forEach(title => {
        if (article.title === title) acc.push(article);
      });
      return acc;
    }, []);
    return favoriteArticles.map(article => (
      <Article key={article.id} article={article} match={match} />
    ));
  };

  const articles = generateArticleCategory();
  const disabledClass = isDisabled ? '--disabled' : '';
  const breakpoints = {
    default: 3,
    1500: 2,
    950: 1
  };

  return articles.length ? (
    <Masonry
      breakpointCols={breakpoints}
      className={'ArticleContainer' + disabledClass}
      columnClassName="columns"
    >
      {articles}
    </Masonry>
  ) : (
    <h1 className="ArticleContainer--no-faves">No favorites yet!</h1>
  );
};

export const mapStateToProps = state => ({
  natGeoArticles: state.natGeoArticles,
  newScientistArticles: state.newScientistArticles,
  cryptoCoinsArticles: state.cryptoCoinsArticles
});

export default connect(mapStateToProps)(ArticleContainer);

ArticleContainer.propTypes = {
  match: PropTypes.object,
  natGeoArticles: PropTypes.array,
  newScientistArticles: PropTypes.array,
  cryptoCoinsArticles: PropTypes.array,
  isDisabled: PropTypes.bool
};
