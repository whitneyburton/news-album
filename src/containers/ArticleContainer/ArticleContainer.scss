.ArticleContainer {
  display: flex;
  flex-wrap: wrap;
  margin-left: -30px;
  // width: auto;

  .columns {
    padding-left: 30px; 
    background-clip: padding-box;
  }
   
  .columns > div {
    background: grey;
    margin-bottom: 30px;
  }
}