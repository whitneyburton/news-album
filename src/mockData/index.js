export const mockNatGeoArticle = {
  author: 'Douglas Main',
  title: 'Why insect populations are plummeting',
  description: 'Sad insects.',
  url:
    'https://www.nationalgeographic.com/animals/2019/02/why-insect-populations-are-plummeting-and-why-it-matters.html',
  urlToImage:
    'https://www.nationalgeographic.com/content/dam/animals/2019/02/insects-in-decline/01-insects-in-decline-nationalgeographic_1353541.jpg',
  publishedAt: '2019-02-19T02:07:15.6368398Z',
  id: 'b',
  isFavorite: false
};

export const mockCryptoCoinsArticle = {
  author: 'Greg Thompson',
  title: 'Ethereum Slayers',
  description: 'They are everywhere.',
  url: 'https://www.ccn.com/ethereum-killers-crypto-blockchain',
  urlToImage:
    'https://www.ccn.com/wp-content/uploads/2019/02/ethereum-crypto-dark-cement-shutterstock.jpg',
  publishedAt: '2019-02-23T14:28:21Z',
  id: 'f',
  isFavorite: false
};

export const mockNewScientistArticle = {
  author: 'Leah Crane',
  title: 'Four stars full of burnt nuclear ash',
  description: 'Scary stars!',
  url:
    'https://www.newscientist.com/article/2194683-four-stars-full-of-burnt-nuclear-ash-are-speeding-through-the-galaxy/',
  urlToImage:
    'https://images.newscientist.com/wp-content/uploads/2019/02/22150921/vennes1hr.jpg',
  publishedAt: '2019-02-22T00:00:00Z',
  id: 'd',
  isFavorite: false
};

export const mockNatGeoArticles = [
  {
    author: 'Erica Johnson',
    title: '20 U.S. state parks',
    description: 'Pretty parks!',
    url:
      'https://www.nationalgeographic.com/travel/destinations/north-america/united-states/state-parks-worth-the-trip.html',
    urlToImage:
      'https://www.nationalgeographic.com/content/dam/travel/2019-digital/state-parks-to-know/02-state-parks-waianapanapa-park-hawaii.jpg',
    publishedAt: '2019-02-19T02:07:16.9805881Z',
    id: 'a',
    isFavorite: false
  },
  {
    author: 'Douglas Main',
    title: 'Why insect populations are plummeting',
    description: 'Sad insects.',
    url:
      'https://www.nationalgeographic.com/animals/2019/02/why-insect-populations-are-plummeting-and-why-it-matters.html',
    urlToImage:
      'https://www.nationalgeographic.com/content/dam/animals/2019/02/insects-in-decline/01-insects-in-decline-nationalgeographic_1353541.jpg',
    publishedAt: '2019-02-19T02:07:15.6368398Z',
    id: 'b',
    isFavorite: false
  }
];

export const mockNatGeoArticlesWithFav = [
  {
    author: 'Erica Johnson',
    title: '20 U.S. state parks',
    description: 'Pretty parks!',
    url:
      'https://www.nationalgeographic.com/travel/destinations/north-america/united-states/state-parks-worth-the-trip.html',
    urlToImage:
      'https://www.nationalgeographic.com/content/dam/travel/2019-digital/state-parks-to-know/02-state-parks-waianapanapa-park-hawaii.jpg',
    publishedAt: '2019-02-19T02:07:16.9805881Z',
    id: 'a',
    isFavorite: false
  },
  {
    author: 'Douglas Main',
    title: 'Why insect populations are plummeting',
    description: 'Sad insects.',
    url:
      'https://www.nationalgeographic.com/animals/2019/02/why-insect-populations-are-plummeting-and-why-it-matters.html',
    urlToImage:
      'https://www.nationalgeographic.com/content/dam/animals/2019/02/insects-in-decline/01-insects-in-decline-nationalgeographic_1353541.jpg',
    publishedAt: '2019-02-19T02:07:15.6368398Z',
    id: 'b',
    isFavorite: true
  }
];

export const mockNewScienceArticles = [
  {
    author: 'Chelsea Whyte',
    title: 'Wolves are spreading in Germany',
    description: 'Wolves everywhere!',
    url:
      'https://www.nationalgeographic.https://www.newscientist.com/article/2194696-wolves-are-spreading-in-germany-by-moving-into-military-land//animals/2019/02/why-insect-populations-are-plummeting-and-why-it-matters.html',
    urlToImage:
      'https://images.newscientist.com/wp-content/uploads/2019/02/22145340/rm7694.jpg',
    publishedAt: '2019-02-23T00:00:00Z',
    id: 'c',
    isFavorite: false
  },
  {
    author: 'Leah Crane',
    title: 'Four stars full of burnt nuclear ash',
    description: 'Scary stars!',
    url:
      'https://www.newscientist.com/article/2194683-four-stars-full-of-burnt-nuclear-ash-are-speeding-through-the-galaxy/',
    urlToImage:
      'https://images.newscientist.com/wp-content/uploads/2019/02/22150921/vennes1hr.jpg',
    publishedAt: '2019-02-22T00:00:00Z',
    id: 'd',
    isFavorite: false
  }
];

export const mockNewScienceArticlesWithFav = [
  {
    author: 'Chelsea Whyte',
    title: 'Wolves are spreading in Germany',
    description: 'Wolves everywhere!',
    url:
      'https://www.nationalgeographic.https://www.newscientist.com/article/2194696-wolves-are-spreading-in-germany-by-moving-into-military-land//animals/2019/02/why-insect-populations-are-plummeting-and-why-it-matters.html',
    urlToImage:
      'https://images.newscientist.com/wp-content/uploads/2019/02/22145340/rm7694.jpg',
    publishedAt: '2019-02-23T00:00:00Z',
    id: 'c',
    isFavorite: false
  },
  {
    author: 'Leah Crane',
    title: 'Four stars full of burnt nuclear ash',
    description: 'Scary stars!',
    url:
      'https://www.newscientist.com/article/2194683-four-stars-full-of-burnt-nuclear-ash-are-speeding-through-the-galaxy/',
    urlToImage:
      'https://images.newscientist.com/wp-content/uploads/2019/02/22150921/vennes1hr.jpg',
    publishedAt: '2019-02-22T00:00:00Z',
    id: 'd',
    isFavorite: true
  }
];

export const mockCryptoCoinsArticles = [
  {
    author: 'Joseph Young',
    title: 'Bitcoins spike',
    description: 'So much money!',
    url:
      'https://www.ccn.com/bitcoin-price-4100-2-hours-analyst-crypto-bullish',
    urlToImage:
      'https://www.ccn.com/wp-content/uploads/2019/02/bitcoin-price-green-chart-bull-shutterstock.jpg',
    publishedAt: '2019-02-23T14:45:23Z',
    id: 'e',
    isFavorite: false
  },
  {
    author: 'Greg Thompson',
    title: 'Ethereum Slayers',
    description: 'They are everywhere.',
    url: 'https://www.ccn.com/ethereum-killers-crypto-blockchain',
    urlToImage:
      'https://www.ccn.com/wp-content/uploads/2019/02/ethereum-crypto-dark-cement-shutterstock.jpg',
    publishedAt: '2019-02-23T14:28:21Z',
    id: 'f',
    isFavorite: false
  }
];

export const mockCryptoCoinsArticlesWithFav = [
  {
    author: 'Joseph Young',
    title: 'Bitcoins spike',
    description: 'So much money!',
    url:
      'https://www.ccn.com/bitcoin-price-4100-2-hours-analyst-crypto-bullish',
    urlToImage:
      'https://www.ccn.com/wp-content/uploads/2019/02/bitcoin-price-green-chart-bull-shutterstock.jpg',
    publishedAt: '2019-02-23T14:45:23Z',
    id: 'e',
    isFavorite: false
  },
  {
    author: 'Greg Thompson',
    title: 'Ethereum Slayers',
    description: 'They are everywhere.',
    url: 'https://www.ccn.com/ethereum-killers-crypto-blockchain',
    urlToImage:
      'https://www.ccn.com/wp-content/uploads/2019/02/ethereum-crypto-dark-cement-shutterstock.jpg',
    publishedAt: '2019-02-23T14:28:21Z',
    id: 'f',
    isFavorite: true
  }
];

export const mockNatGeoArticlesObj = {
  articles: [
    {
      author: 'Erica Johnson',
      title: '20 U.S. state parks',
      description: 'Pretty parks!',
      url:
        'https://www.nationalgeographic.com/travel/destinations/north-america/united-states/state-parks-worth-the-trip.html',
      urlToImage:
        'https://www.nationalgeographic.com/content/dam/travel/2019-digital/state-parks-to-know/02-state-parks-waianapanapa-park-hawaii.jpg',
      publishedAt: '2019-02-19T02:07:16.9805881Z',
      id: 'a',
      isFavorite: false
    },
    {
      author: 'Douglas Main',
      title: 'Why insect populations are plummeting',
      description: 'Sad insects.',
      url:
        'https://www.nationalgeographic.com/animals/2019/02/why-insect-populations-are-plummeting-and-why-it-matters.html',
      urlToImage:
        'https://www.nationalgeographic.com/content/dam/animals/2019/02/insects-in-decline/01-insects-in-decline-nationalgeographic_1353541.jpg',
      publishedAt: '2019-02-19T02:07:15.6368398Z',
      id: 'b',
      isFavorite: false
    }
  ]
};

export const mockNewScienceArticlesObj = {
  articles: [
    {
      author: 'Chelsea Whyte',
      title: 'Wolves are spreading in Germany',
      description: 'Wolves everywhere!',
      url:
        'https://www.nationalgeographic.https://www.newscientist.com/article/2194696-wolves-are-spreading-in-germany-by-moving-into-military-land//animals/2019/02/why-insect-populations-are-plummeting-and-why-it-matters.html',
      urlToImage:
        'https://images.newscientist.com/wp-content/uploads/2019/02/22145340/rm7694.jpg',
      publishedAt: '2019-02-23T00:00:00Z',
      id: 'c',
      isFavorite: false
    },
    {
      author: 'Leah Crane',
      title: 'Four stars full of burnt nuclear ash',
      description: 'Scary stars!',
      url:
        'https://www.newscientist.com/article/2194683-four-stars-full-of-burnt-nuclear-ash-are-speeding-through-the-galaxy/',
      urlToImage:
        'https://images.newscientist.com/wp-content/uploads/2019/02/22150921/vennes1hr.jpg',
      publishedAt: '2019-02-22T00:00:00Z',
      id: 'd',
      isFavorite: false
    }
  ]
};

export const mockCryptoCoinsArticlesObj = {
  articles: [
    {
      author: 'Joseph Young',
      title: 'Bitcoins spike',
      description: 'So much money!',
      url:
        'https://www.ccn.com/bitcoin-price-4100-2-hours-analyst-crypto-bullish',
      urlToImage:
        'https://www.ccn.com/wp-content/uploads/2019/02/bitcoin-price-green-chart-bull-shutterstock.jpg',
      publishedAt: '2019-02-23T14:45:23Z',
      id: 'e',
      isFavorite: false
    },
    {
      author: 'Greg Thompson',
      title: 'Ethereum Slayers',
      description: 'They are everywhere.',
      url: 'https://www.ccn.com/ethereum-killers-crypto-blockchain',
      urlToImage:
        'https://www.ccn.com/wp-content/uploads/2019/02/ethereum-crypto-dark-cement-shutterstock.jpg',
      publishedAt: '2019-02-23T14:28:21Z',
      id: 'f',
      isFavorite: false
    }
  ]
};

export const mockNewSciArticlesStringified = {
  json: {
    articles: [
      {
        author: 'Chelsea Whyte',
        description: 'Wolves everywhere!',
        id: 'c',
        isFavorite: false,
        publishedAt: '2019-02-23T00:00:00Z',
        title: 'Wolves are spreading in Germany',
        url:
          'https://www.nationalgeographic.https://www.newscientist.com/article/2194696-wolves-are-spreading-in-germany-by-moving-into-military-land//animals/2019/02/why-insect-populations-are-plummeting-and-why-it-matters.html',
        urlToImage:
          'https://images.newscientist.com/wp-content/uploads/2019/02/22145340/rm7694.jpg'
      },
      {
        author: 'Leah Crane',
        description: 'Scary stars!',
        id: 'd',
        isFavorite: false,
        publishedAt: '2019-02-22T00:00:00Z',
        title: 'Four stars full of burnt nuclear ash',
        url:
          'https://www.newscientist.com/article/2194683-four-stars-full-of-burnt-nuclear-ash-are-speeding-through-the-galaxy/',
        urlToImage:
          'https://images.newscientist.com/wp-content/uploads/2019/02/22150921/vennes1hr.jpg'
      }
    ]
  },
  ok: true
};
