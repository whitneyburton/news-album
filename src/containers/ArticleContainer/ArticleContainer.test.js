import React from 'react';
import { shallow } from 'enzyme';
import { ArticleContainer, mapStateToProps } from './ArticleContainer';
import * as data from '../../mockData';

const mockProps = {
  match: { path: '/national-geographic' },
  natGeoArticles: data.mockNatGeoArticles,
  newScientistArticles: data.mockNewScienceArticles,
  cryptoCoinsArticles: data.mockCryptoCoinsArticles,
  isDisabled: false
};

describe('ArticleContainer', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ArticleContainer {...mockProps} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  // describe('generateArticleCategory', () => {
  //   it('should return an array of Article components when there is a NAT GEO path', () => {
  //     const result = wrapper.instance().generateArticleCategory();
  //     expect(result).toHaveLength(2);
  //   });

  //   it('should return an array of Article components when there is a CRYPTO COINS path', () => {
  //     const mockMatch = { path: '/national-geographic' };
  //     wrapper = shallow(<ArticleContainer {...mockProps} match={mockMatch} />);
  //     const result = wrapper.instance().generateArticleCategory();
  //     expect(result).toHaveLength(2);
  //   });

  //   it('should return an array of Article components when there is a NEW SCIENTIST path', () => {
  //     const mockMatch = { path: '/new-scientist' };
  //     wrapper = shallow(<ArticleContainer {...mockProps} match={mockMatch} />);
  //     const result = wrapper.instance().generateArticleCategory();
  //     expect(result).toHaveLength(2);
  //   });

  //   it('should return an array of Article components when there is a FAVORITES path', () => {
  //     const mockMatch = { path: '/favorites' };
  //     wrapper = shallow(<ArticleContainer {...mockProps} match={mockMatch} />);
  //     const result = wrapper.instance().generateArticleCategory();
  //     expect(result).toHaveLength(0);
  //   });
  // });

  describe('mapStateToProps', () => {
    it('should return a props object with all 3 types of articles', () => {
      const mockState = {
        natGeoArticles: data.mockNatGeoArticles,
        newScientistArticles: data.mockNewScientistArticle,
        cryptoCoinsArticles: data.mockCryptoCoinsArticles,
        isLoading: false
      };
      const expected = {
        natGeoArticles: data.mockNatGeoArticles,
        newScientistArticles: data.mockNewScientistArticle,
        cryptoCoinsArticles: data.mockCryptoCoinsArticles
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });
});
