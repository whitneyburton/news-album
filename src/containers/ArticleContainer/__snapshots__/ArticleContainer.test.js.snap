// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleContainer should match the snapshot 1`] = `
<h1
  className="ArticleContainer--no-faves"
>
  No favorites yet!
</h1>
`;
